### DISCUSSION TITLE
SS2K Overlay

### DISCUSSION THREAD

**Discussion by MarkusSchneider** [doudar/SmartSpin2k#286] [Ideas]
I've create an electron app to show smart spin 2k data during workout.
https://github.com/MarkusSchneider/SmartSpin2kOverlay
The data is collected via UDP (from the logs).

@doudar if you think this is interesting, you could add it to the https://github.com/SmartSpin2k repository and we can work together on the app.
We can also display any other data like Shifter step / Gear in simulation mode or what every we want to collect.

![image](https://user-images.githubusercontent.com/7011066/147491744-8d107e2d-d5b0-4c6f-a305-2ea79660f63e.png)

Any feedback is welcome.

Greetings
Markus

**Comment by doudar** [comment]
That's actually pretty cool! 

Would there be a way to show connected devices and perhaps the power table as well? 

How about future ability to change settings and other things from the ui?

**Reply by MarkusSchneider** [reply]
We can show there everything that we get out of the logs, also connected Devices and power table.
It should be possible to also change settings from the electron app. CORS could be a problem but could be solved.

Electron works on Windows, Linux and MacOS if you ran Zwift on Tablet it's not a good solution.

**Reply by doudar** [reply]
I run Zwift on Windows. Not sure what the userbase primarily uses, but at a minimum (SS2K Overlay) seems like a great and easy diagnostic tool. 

I've never built anything with Electron, but I looked over the code and it seems to be all pretty easy Javascript, so that's very nice.

After doing some research on it today, would it be worthwhile (or feasible) to think about using [Apache Cordova](https://cordova.apache.org/) for this? I've never worked with it either, but it seems similar to Electron and if it did everything you need it to, it *might* allow a same code base native SS2K settings app for Windows, Mac, iOS and Android once we built out all the features. Edit: and maybe Cordova is getting outdated?

If Electron or Cordova supports BLE, we might have the ability to do some of this via the custom characteristic vv. 

https://github.com/doudar/SmartSpin2k/blob/fedebad867293ae03e58b4af782b4ac7ee1996df/src/BLE_Server.cpp#L500-L811

**Reply by MarkusSchneider** [reply]
For the overlay I must be able to create a window on top of Zwift without border , menu, frames, ...
I also never worked with Cordova and don't know if this is possible.
State of the art is to create a [progressiv web app](https://en.wikipedia.org/wiki/Progressive_web_application) to run your code native.
PWA is more limited in the interaction with the OS but its maybe not a huge problem.
I think I'll try this next.
I like the idea to have all important values on one screen 😁.

BEL could work with [nobel](https://github.com/noble/noble) and electron, for PWA I think its not implemented.

**Reply by doudar** [reply]
Yes, PWA Is probably the best way. I think they're all competing for the same space and it was hard for me to tell which has the most support and user base with limited research. 

Will PWA allow the always on top window in windows Zwift that you like?

**Reply by doudar** [reply]
Possibly we could create a TCP socket on the SS2K as well. I'm not certain if that causes issues with the BLE and it would need to be tested.

**Comment by MarkusSchneider** [comment]
Not sure it a PWA would stay always on top. A PWA is more or less a web application running in the browser. But a simple workaround would be to open zwift and move the pwa-window over the zwift screen.
More important for me is the possibility to hide the title bar of the browser. 

Yes TCP socket would maybe be the best solution or a web socket, but I saw the discussion to integrate AsyncWebServer an the problems  😒. A HTTP based communication would be nice because its supported well by browsers.

**Reply by doudar** [reply]
We could always try testing it (asynchronous web server) again . There's been a lot of changes in the last year. @h2zero , do you know if NimBLE and async are working well together nowadays?

**Reply by doudar** [reply]
Also, I was thinking you could use  the existing http get method within the app to get the config data via JSON or make a new http handler that sends the JSON via debug output (udp).

**Reply by h2zero** [reply]
I haven't done any recent testing with async + BLE might be time to try it again.

**Reply by h2zero** [reply]
I think I have an answer: https://github.com/h2zero/NimBLE-Arduino/issues/323#issuecomment-1002236218

**Reply by doudar** [reply]
> I think I have an answer: https://github.com/h2zero/NimBLE-Arduino/issues/323#issuecomment-1002236218

Dang! There it is again 😅
